### ISSUE TITLE
testVersion: automatically detect package

### ISSUE THREAD

**Issue by davidak** [NixOS/nixpkgs#190577]
## Issue description

Currently the `package` name has to be set manually every time.

It would be great if `testVersion` can detect it somehow, so you don't have to repeat it (DRY).

Then a minimal config. looks like this:

```
 passthru.tests.version = testers.testVersion {}; 
```

There are some packages in nixpkgs that use some form where they don't have to write the package name, but my naive approach to just copy it to `testVersion`s code does not work.

https://github.com/NixOS/nixpkgs/blob/350fd0044447ae8712392c6b212a18bdf2433e71/pkgs/development/tools/amazon-qldb-shell/default.nix#L33

https://github.com/NixOS/nixpkgs/blob/d14d17e0f103ab03cd9fbe0f883569f19c36f6e7/pkgs/applications/misc/swaynotificationcenter/default.nix#L75-L78

As described in https://nixos.org/manual/nixpkgs/stable/#var-meta-tests.

https://github.com/NixOS/nixpkgs/blob/350fd0044447ae8712392c6b212a18bdf2433e71/pkgs/development/compilers/go-jsonnet/default.nix#L20-L23

cc @raboof @roberth  @0x4A6F @Artturin

Is there any chance this is possible with Nix?

**Comment by roberth** [comment]
> passthru.tests.version = testers.testVersion {};

The purpose of `passthru` is to humbly set attributes on the package attrset, so unless we change it to do something it wasn't designed to do, it shouldn't work, because the expression `testers.testVersion {}` has no reference to the package.

I can imagine a module-system-like module that can express your intent even more concisely, but for now we don't have such a feature in `mkDerivation`.

**Comment by davidak** [comment]
Thanks for the response.

> because the expression `testers.testVersion {}` has no reference to the package.

When thinking naively without knowing anything about Nix internals, i think Nix can be able to know for which package a test is when you run them with `nix build .#vapoursynth.tests.version`. Also, the code is in the same file and even code block as the package. I think it should be possible somehow technically, but maybe not with the current architecture.

The architecture should enable us to write readable, elegant and short code and not limit us in that goal.

**Comment by roberth** [comment]
I agree with the ambition that it should be simple to specify a test, but by removing the need for `inherit finalPackage`, you move the complexity from the package, where it can be easily understood, into the interface, where it is magical.

This is why I brought up module style composition, as a different kind of interface that better supports the kind of logic you want.

> The architecture should enable us to write readable, elegant and short code and not limit us in that goal.

Decent goals that need to be balanced by the same goals at the other side of the interface.
